### PR TITLE
MISRA-C Checking function return

### PR DESCRIPTION
**Pull request by ceolin**
This pr is part of #9884 feature. There still violations that will be addressed in others commits/pr.

Important points about this pr:

- Commit changing signature of __swap()
  - This function can return -EAGAIN but its return value was unsigned int. Nevertheless, this value was being propagate as int.
- Change printk()/vnprintk() signature to void

### CODE REVIEW

**Comment by pfalcon** [comment]
To the best of my knowledge, the MISRA C experiment (or whatever the right name) should have been applied to "kernel/*". This is explicitly stated in https://github.com/zephyrproject-rtos/zephyr/issues/9552

It would be nice to indeed follow that schedule, or otherwise we get these patches with 237 changed files, attempt to looking thru which shows memset, memset... Again, #9552 states "Adhere to MISRA-C 2012 where it makes sense and provide deviations and documentation of rules", and one of obvious solution for memset is not patching it all the way thru, but leaving alone. it would be nice to see that choice be followed/discussed too.

**Review by jhedberg — CHANGES_REQUESTED** [review]
Doing a void cast on memset seems rather pointless to me, i.e. sounds like a good exception to be made from the MISRA rules. The return value of memset does not exist to denote error/success (which would be a valid reason to acknowledge it when doing the call) rather it's always the same thing, i.e. a pointer to the given memory area, which is rarely useful.

**Review by jukkar — CHANGES_REQUESTED** [review]
I agree with @jhedberg and @pfalcon here. At this first stage we should only concentrate changes in the kernel code. Also changing memset() calls seems quite pointless.

**Review by andyross — APPROVED** [review]
OK, these rules aren't awful.  Frankly most of those (void)'s are needed because our own poor API choices anyway.

**Comment by andyross** [comment]
And just to jump in to the bike shed: the memset() stuff doesn't bother me personally.  The MISRA rule here is actually a very sane one to enforce in the general case, and it's not the fault of the MISRA standard that the C standard has a semi-broken API that returns a value needlessly.

That said, surely we could get an exception for whitelisting it though, right?   There are a few other APIs with "optional but rarely used" return values (printf comes immediately to mind) where the same argument would apply.

**Comment by ceolin** [comment]
@jhedberg @pfalcon It's not a matter or like or not this. The fact is, this is required guideline, which means that we have to adhere this or create a satisfactory deviation. I personally don't like as well, but it does not hurt at least. APIs we can change, I'm totally open to discuss and see the best we can do.

If we want to be compliant, we'll need to do such things, this guideline explicitly says that every non-void functions must have its return checked, if we don't do anything with memset() return, we need to explicitly do the (void) to acknowledge that.

About, stick only with kernel/*, I can do it, no problem. Actually more disruptive changes I'm doing it, but for these harmless things I was applying for whole project, so in the future when we decide to apply the guideline to whole project we'll have less work to do.

**Comment by ceolin** [comment]
One more point, create deviations make harder to show compliance, AFAIU. So, for me we should create deviations that are really necessary.

File: kernel/timer.c [hunk 1]
```diff
@@ -123,7 +123,10 @@ void _impl_k_timer_start(struct k_timer *timer, s32_t duration, s32_t period)
 	unsigned int key = irq_lock();
 
 	if (timer->timeout.delta_ticks_from_prev != _INACTIVE) {
-		_abort_timeout(&timer->timeout);
+		/* _abort_timeout() will return alway 0 since the
+		   condition to return _INACTIVE was checked.
+		 */
+		(void)_abort_timeout(&timer->timeout);
```
[the remark this answers is not in the document]
Comment (ceolin) [reply kernel/timer.c]: yeah, I was thinking about it too. What is better, from performance perspective, always do the call and let the check there, or avoid unecessary call doing the check. As I don't have a clue which case is more common, I just left the comment. I'm open to change it (if you know what is better) or remove.

[the remark this answers is not in the document]
Comment (andyross) [reply kernel/timer.c]: _abort_timeout() is an inline, so the generated code is almost certainly going to be the same either way.  Personally I'd remove the test, but it's a minor aesthetic thing and not worth arguing about.

**Comment by nashif** [comment]
add (void) to a function that returns something signifies that we know the functions returns something but we will not be checking it. This way, someone looking at the code or when the code is being audited it is immediately clear that we did not forget to check the return.

How many coverity issues related to not checking return values from function did you all fix? 

It looks odd and "not needed" etc etc, but this is the sort of thing that adds the look of safety to something that could already have been safe.. 

We can go wild create deviations but I prefer that we do that for the really tricky items and not the trivial ones.

**Comment by ceolin** [comment]
@nashif completely agree, we should avoid deviations as much as possible. Regarding the number of similar violations in kernel/*, we have ~60 more.
I'm treating each case individually, seeing if makes sense change API, if we should treat the return, if we can ignore ...

This memset was massive because we can't change this API and because I applied to the whole project because this one don't have different use cases. The others are less disruptive.

File: kernel/timer.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (ceolin) [reply kernel/timer.c]: it'll be done :)

**Review by jukkar — APPROVED** [review]

**Comment by jhedberg** [comment]
I guess we'll going to have to agree to disagree on whether this memset cast actually makes the code any more safe or secure :) Btw, has anyone considered using `__attribute__((warn_unused_result))` instead to improve the code quality? At least both gcc and llvm supports it.

**Comment by nashif** [comment]
> I guess we're going to have to agree to disagree on whether this memset cast actually makes the code any more safe or secure :)

nobody is saying that. Following a coding standard or guideline does not make the code safer, and I know it looks really bad with all the voids for memset. This is analog to why we have a coding style and why we format the code the way we do, more or less and provides a standard and consistent way for coding that help spot issues and know the original intention of the developer who wrote the code without too much guessing.

**Comment by jhedberg** [comment]
@nashif understood, but I highly doubt we'd be coming up with this convention (or being blindly strict about it) if it wasn't for MISRA-C

**Review by jhedberg — APPROVED** [review]
Approving, but if CI doesn't start blocking new PRs which don't follow these conventions we'll soon again have violations of them in the tree. E.g. I'm aware of at least one pending PR that adds a memset (without the cast).

**Review by nashif — APPROVED** [review]

**Comment by ceolin** [comment]
@jhedberg #9884 contains more information about it, but yeah, I'll add __attribute__((warn_unused_result)) in functions that must have their return checked.

**Comment by ceolin** [comment]
just one more point about using `attribute((warn_unused_result))`. When we add this for a function we can't ignore its result doing `(void)`, so we can use this only for the functions that we *really* should handle their return.

**Comment by ceolin** [comment]
Added one more commit with leftovers. Now, all these cases on kernel are fixed.

**Comment by ceolin** [comment]
@nashif can we merge it ?

**Pull request merged**